graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java

```diff
@@ -0,0 +1,178 @@
+    }
+
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @ApiOperation("Create a new entity group")
+    @ApiResponses(@ApiResponse(code = 400, message = "An entity group already exists with id or name"))
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_EDIT)
+    @AuditEvent(type = AuditEventTypes.ENTITY_GROUP_CREATE)
+    public Response create(@ApiParam(name = "JSON Body") EntityGroup entityGroup) {
+        try {
+            return Response.ok().entity(entityGroupService.create(entityGroup)).build();
+        } catch (DuplicateKeyException | MongoWriteException e) {
+            throw new BadRequestException(StringUtils.f("Entity group '%s' already exists", entityGroup.name()));
+        }
+    }
+
+    @PUT
+    @Path("/{id}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @ApiOperation("Update an entity group")
+    @ApiResponses({
+            @ApiResponse(code = 400, message = "An entity group already exists with id or name"),
+            @ApiResponse(code = 404, message = "No such entity group")
+    })
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_EDIT)
+    @AuditEvent(type = AuditEventTypes.ENTITY_GROUP_UPDATE)
+    public Response update(@ApiParam(name = "id", required = true) @PathParam("id") String id,
+                           @ApiParam(name = "JSON Body") EntityGroup entityGroup) {
+        try {
+            return Response.ok().entity(entityGroupService.update(id, entityGroup)).build();
+        } catch (IllegalArgumentException e) {
+            throw new NotFoundException(getRootCauseMessage(e), e);
+        } catch (DuplicateKeyException | MongoWriteException e) {
+            throw new BadRequestException(StringUtils.f("Entity group '%s' already exists", entityGroup.name()));
+        }
+    }
+
+    @PUT
+    @Path("/{group_id}/add_entity")
+    @ApiOperation("Add an entity to an entity group")
+    @ApiResponses(@ApiResponse(code = 404, message = "No such entity group"))
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_EDIT)
+    @AuditEvent(type = AuditEventTypes.ENTITY_GROUP_UPDATE)
+    public Response addEntityToGroup(@ApiParam(name = "group_id", required = true) @PathParam("group_id") String groupId,
+                                     @ApiParam(name = "entity_type", required = true) @QueryParam("entity_type") String entityType,
+                                     @ApiParam(name = "entity_id", required = true) @QueryParam("entity_id") String entityId) {
+        try {
+            return Response.ok().entity(entityGroupService.addEntityToGroup(groupId, entityType, entityId)).build();
+        } catch (IllegalArgumentException e) {
+            throw new NotFoundException(getRootCauseMessage(e), e);
+        }
+    }
+
+    @PUT
+    @Path("/get_for_entities")
+    @ApiOperation("Get a list of entity groups for a list of entities")
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_READ)
```

Comment:
Permission checks on resources that deal with a specific instance should use per-instance checks, not only annotations. Compare eg. https://github.com/Graylog2/graylog2-server/blob/e53295c35322939ad25a4d3eb67196604bd5d321/graylog2-server/src/main/java/org/graylog2/rest/resources/users/UsersResource.java#L191